8321206: Make Locale related system properties `StaticProperty`

naotoj opened the pull request:
Currently, Locale-related system properties, such as `user.language` or `user.country`, are initialized when the `Locale` class is loaded. Making them static properties is safer than relying on the `Locale` class loading timing, which could potentially be changed depending on the implementation.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321206](https://bugs.openjdk.org/browse/JDK-8321206): Make Locale related system properties `StaticProperty` (**Sub-task** - P4)


### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16986/head:pull/16986` \
`$ git checkout pull/16986`

Update a local copy of the PR: \
`$ git checkout pull/16986` \
`$ git pull https://git.openjdk.org/jdk.git pull/16986/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16986`

View PR using the GUI difftool: \
`$ git pr show -t 16986`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16986.diff">https://git.openjdk.org/jdk/pull/16986.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16986#issuecomment-1841774355)

dholmes-ora commented:
I'm not following the changes to cdsHeapVerifier.cpp. You've marked the new entries as `C` but the definition is:
```
// [C] A non-final static string that is assigned a string literal during class
// initialization; this string is never changed during -Xshare:dump.
```
and these are final static strings not non-final. ???

dholmes-ora commented:
Can I also suggest that you change the bug and PR synopsis to say `StaticProperty` rather than `static properties` as I found it quite confusing to understand the issue.

> Making them static properties is safer than relying on the class initialization timing

"class initialization" refers to the static initialization of a class. I assume that is not what you mean here but the creation of the instance of the class? Even `StaticProperty` still initializes these things during class initialization, so I'm unclear exactly what this is trying to say, or the the change is wanted/desirable.

AlanBateman commented on src/java.base/share/classes/java/util/Locale.java:
```diff
@@ -1069,30 +1074,29 @@ private static Locale initDefault() {
+            StaticProperty.userScript(category.ordinal() + 1),
+            StaticProperty.userCountry(category.ordinal() + 1),
+            StaticProperty.userVariant(category.ordinal() + 1),
+            getDefaultExtensions(StaticProperty.userExtensions(category.ordinal() + 1))
```
I suspect this would be more readable, and maintainable, if you add non-public languageKey/scriptKey/countryKey/variantKey methods to Locale.Category. That would change this to getInstance(category.languageKey(), ...) so no sight of the ordinal at the use-site.

RogerRiggs commented on src/java.base/share/classes/jdk/internal/util/StaticProperty.java:
```diff
@@ -276,4 +308,109 @@ public static String osArch() {
+     *                 {@code Locale.Category.ordinal() + 1} for the category
+     *                 specific property
+     */
+    public static String userVariant(int category) {
+        return switch (category) {
+            case 0 -> USER_VARIANT;
+            case 1 -> USER_VARIANT_DISPLAY;
+            case 2 -> USER_VARIANT_FORMAT;
+            default -> throw new InternalError();
+        };
+    }
+
+    /**
+     * {@return the {@code user.extensions} system property}
+     * <strong>{@link SecurityManager#checkPropertyAccess} is NOT checked
+     * in this method. The caller of this method should take care to ensure
+     * that the returned property is not made accessible to untrusted code.</strong>
+     *
+     * @param category locale category. 0 for the base property,
+     *                 {@code Locale.Category.ordinal() + 1} for the category
+     *                 specific property
+     */
+    public static String userExtensions(int category) {
+        return switch (category) {
+            case 0 -> USER_EXTENSIONS;
+            case 1 -> USER_EXTENSIONS_DISPLAY;
+            case 2 -> USER_EXTENSIONS_FORMAT;
+            default -> throw new InternalError();
+        };
+    }
+
+    /**
+     * {@return the {@code user.region} system property}
+     * <strong>{@link SecurityManager#checkPropertyAccess} is NOT checked
+     * in this method. The caller of this method should take care to ensure
+     * that the returned property is not made accessible to untrusted code.</strong>
+     */
+    public static String userRegion() {
+        return USER_REGION;
+    }
```
Using methods to retrieve these makes is more complicated.
The bleeding of the enum values outside of Locale is undesirable.
Since the property values are final strings, I suggest just making the fields public and keep the mapping local to the Locale class.

naotoj replied:
As Alan commented, I will not use `Locale.Category.ordinal()` but instead use the properties keys. Would that address your suggestion?

naotoj replied:
Ended up replacing public methods to public fields

RogerRiggs commented on src/java.base/share/classes/java/util/Locale.java:
```diff
@@ -1051,13 +1052,17 @@ private static synchronized Locale getFormatLocale() {
         String language, region, script, country, variant;
-        Properties props = GetPropertyAction.privilegedGetProperties();
-        language = props.getProperty("user.language", "en");
+        var sm = System.getSecurityManager();
+        if (sm != null) {
+            sm.checkPropertiesAccess();
+        }
```
This SM check is no longer needed; the use of privilegedGetProperties made access unconditional.
The static properties are always accessible to the locale implementation.

naotoj replied:
Was wondering so. Will remove them.

RogerRiggs commented on src/java.base/share/classes/jdk/internal/util/StaticProperty.java:
```diff
@@ -79,6 +95,22 @@ private StaticProperty() {}
         OS_NAME = getProperty(props, "os.name");
         OS_ARCH = getProperty(props, "os.arch");
         OS_VERSION = getProperty(props, "os.version");
+        USER_LANGUAGE = getProperty(props, "user.language", "en");
+        USER_LANGUAGE_DISPLAY = getProperty(props, "user.language.display", USER_LANGUAGE);
+        USER_LANGUAGE_FORMAT = getProperty(props, "user.language.format", USER_LANGUAGE);
+        USER_SCRIPT = getProperty(props, "user.script", "");
+        USER_SCRIPT_DISPLAY = getProperty(props, "user.script.display", USER_SCRIPT);
+        USER_SCRIPT_FORMAT = getProperty(props, "user.script.format", USER_SCRIPT);
+        USER_COUNTRY = getProperty(props, "user.country", "");
+        USER_COUNTRY_DISPLAY = getProperty(props, "user.country.display", USER_COUNTRY);
+        USER_COUNTRY_FORMAT = getProperty(props, "user.country.format", USER_COUNTRY);
+        USER_VARIANT = getProperty(props, "user.variant", "");
+        USER_VARIANT_DISPLAY = getProperty(props, "user.variant.display", USER_VARIANT);
+        USER_VARIANT_FORMAT = getProperty(props, "user.variant.format", USER_VARIANT);
+        USER_EXTENSIONS = getProperty(props, "user.extensions", "");
+        USER_EXTENSIONS_DISPLAY = getProperty(props, "user.extensions.display", USER_EXTENSIONS);
+        USER_EXTENSIONS_FORMAT = getProperty(props, "user.extensions.format", USER_EXTENSIONS);
+        USER_REGION = getProperty(props, "user.region", "");
```
Computing the defaults for these properties should be in the Locale class, close to the initialization of the default locale. Splitting the responsibility across files makes it harder to follow what happens where/when.

naotoj replied:
If I am not mistaken, these assignments should stay here in `StaticProperty`, as that is the whole purpose of this change. Depending on `Locale` class loading time might be fragile, and this change guarantees the properties are initialized and frozen in `System.initPhase1()`

naotoj commented:
> I'm not following the changes to cdsHeapVerifier.cpp. You've marked the new entries as `C` but the definition is:
> 
> ```
> // [C] A non-final static string that is assigned a string literal during class
> // initialization; this string is never changed during -Xshare:dump.
> ```
> 
> and these are final static strings not non-final. ???

I simply followed the fix to https://bugs.openjdk.org/browse/JDK-8295295 where the last time I added a `StaticProperty`, it broke the CDS test. Since I am not familiar in this area, I am happy to have the reasoning corrected. I would appreciate suggestions.

> You also have a large number of test failures with this PR.

Are you referring to GHA tests? I am not sure they are relevant to this fix, as my run for mach5 did not cause any regression test failures.

> Can I also suggest that you change the bug and PR synopsis to say StaticProperty rather than static properties as I found it quite confusing to understand the issue.

Modified.

>> Making them static properties is safer than relying on the class initialization timing
> "class initialization" refers to the static initialization of a class. I assume that is not what you mean here but the creation of the instance of the class? Even StaticProperty still initializes these things during class initialization, so I'm unclear exactly what this is trying to say.

Modified it to specifically saying `Locale` class loading time.

naotoj commented:
Thanks for the reviews. I will ask hotspot folks to rectify the reasonings in `cdsHeapVerifier.cpp` if necessary.

/integrate

ecki commented:
This should probably get a release note, i think I remember I saw and even used setProperty in main to set a language locale.

ecki commented:
> in recent JDK releases.

ok, could be that this has changed before. It does make sense to not allow it if the sequence of initialisation can’t be guaranteed.

dholmes-ora commented:
@naotoj  the hotspot changes were not approved by any Hotspot engineer. Unless trivial, all hotspot code changes require two reviews by hotspot developers.

naotoj commented:
Sorry. I forgot the required reviewers' number differs in the hotspot. Let me know if a JBS issue is needed for further modifying the exclusion reasoning.

dholmes-ora commented:
@iklam or @calvinccheung  could either of you take a look at the CDS changes here please. Thanks.

iklam commented:
> @iklam or @calvinccheung could either of you take a look at the CDS changes here please. Thanks.

> I'm not following the changes to cdsHeapVerifier.cpp. You've marked the new entries as C but the definition is:
> ```
> // [C] A non-final static string that is assigned a string literal during class
> // initialization; this string is never changed during -Xshare:dump.
>```
> and these are final static strings not non-final. ???

Hi David, you're correct that the new entries are not of type "C", as the new strings are `static final`. Anyway, the whole reason of this table is to work around the weakness of cdsHeapVerifier, which can sometimes report false positives. The new entries follow the same pattern of the previous entries (which are also wrong ... i.e., `StaticProperty::FILE_ENCODING`)

I have file https://bugs.openjdk.org/browse/JDK-8321940 to improve cdsHeapVerifier. Hopefully I can remove all the entries in cdsHeapVerifier.cpp added by this PR.
